File: google/GoogleQuestion.go
package main

import (
    "fmt"
    "math"
)

const (
    SIZE  = 100
    CARRY = int64(1e10)
)

func main() {
    eFraction := [SIZE]int64{0}
    for i := getIterations(); i > 1; i-- {
        remainder := calculate(&eFraction, 0, i, 1)
        for j := 1; j < SIZE; j++ {
            remainder = calculate(&eFraction, j, i, remainder)
        }
    }

    result := eFraction[0]
    totalPivot := getTotal(result)
    total := 0
    if isPrime(result) {
        fmt.Println(result)
        total = totalPivot
    }

    digits := [10]int{}
    for _, value := range eFraction[1:] {
        for i := 9; i >= 0; i-- {
            digits[i] = int(value % 10)
            value /= 10
        }

        for _, digit := range digits {
            if 0 != total {
                total += digit - int(result / (CARRY / 10))
                if total == totalPivot {
                    result = (result % (CARRY / 10)) * 10 + int64(digit)
                    fmt.Println(result)
                    return
                }
            }

            result = (result % (CARRY / 10)) * 10 + int64(digit)
            if 0 == total && isPrime(result) {
                fmt.Println(result)
                total = totalPivot
            }
        }
    }
}

func getIterations() int {
    digit := SIZE * 10
    result := 1
    remainder := 1.0 / math.E
    for digit >= 0 {
        for remainder < 1 {
            result += 1
            remainder *= float64(result)
        }
        for remainder >= 1 && digit >= 0 {
            remainder /= 10
            digit -= 1
        }
    }
    return result - 1
}

func calculate(eFraction *[SIZE]int64, index int, iteration int, remainder int64) int64 {
    value := eFraction[index] + remainder * CARRY
    eFraction[index] = value / int64(iteration)
    if eFraction[index] >= CARRY {
        eFraction[index] -= CARRY
        eFraction[index - 1]++
    }
    return value % int64(iteration)
}

func getTotal(value int64) int {
    result := 0
    for 0 != value {
        result += int(value % 10)
        value /= 10
    }
    return result
}

func isPrime(value int64) bool {
    if 0 == (value & 1) {
        return 2 == value
    }

    border := int64(math.Sqrt(float64(value)))
    for i := int64(3); i <= border; i += 2 {
        if 0 == value % i {
            return false
        }
    }
    return 1 != value
}

File: google/GoogleQuestion.py
import math
from typing import List

SIZE, CARRY = 100, int(1e10)

def main() -> None:
    e_fraction = [0 for _ in range(SIZE)]
    for i in range(get_iterations(), 1, -1):
        remainder = calculate(e_fraction, 0, i, 1)
        for j in range(1, SIZE):
            remainder = calculate(e_fraction, j, i, remainder)

    result = e_fraction[0]
    total_pivot, total = get_total(result), 0
    if is_prime(result):
        print(result)
        total = total_pivot

    digits = [0 for _ in range(10)]
    for value in e_fraction[1:]:
        for i in range(9, -1, -1):
            digits[i] = value % 10
            value //= 10

        for digit in digits:
            if total:
                total += digit - result // (CARRY // 10)
                if total == total_pivot:
                    result = (result % (CARRY // 10)) * 10 + digit
                    print(result)
                    return

            result = (result % (CARRY // 10)) * 10 + digit
            if not total and is_prime(result):
                print(result)
                total = total_pivot

def get_iterations() -> int:
    digit, result, remainder = SIZE * 10, 1, 1.0 / math.e
    while digit >= 0:
        while remainder < 1:
            result += 1
            remainder *= result
        while remainder >= 1 and digit >= 0:
            remainder /= 10
            digit -= 1
    return result - 1

def calculate(e_fraction: List[int], index: int, iteration: int, remainder: int) -> int:
    value = e_fraction[index] + remainder * CARRY
    e_fraction[index] = value // iteration
    if e_fraction[index] >= CARRY:
        e_fraction[index] -= CARRY
        e_fraction[index - 1] += 1
    return value % iteration

def get_total(value: int) -> int:
    result = 0
    while value:
        result += value % 10
        value //= 10
    return result

def is_prime(value: int) -> bool:
    if not (value & 1):
        return 2 == value

    border = int(math.sqrt(value))
    for i in range(3, border + 1, 2):
        if not value % i:
            return False
    return 1 != value

if __name__ == '__main__':
    main()
